mat/vector.go

```diff
@@ -503,105 +503,118 @@ func (v *VecDense) DivElemVec(a, b Vector) {
 
 // MulVec computes a * b. The result is stored into the receiver.
 // MulVec panics if the number of columns in a does not equal the number of rows in b.
-func (v *VecDense) MulVec(a Matrix, b *VecDense) {
+func (v *VecDense) MulVec(a Matrix, b Vector) {
 	r, c := a.Dims()
-	br := b.Len()
+	br, _ := b.Dims()
 	if c != br {
 		panic(ErrShape)
 	}
 
-	if v != b {
-		v.checkOverlap(b.mat)
+	aU, trans := untranspose(a)
+	var bmat blas64.Vector
+	fast := true
+	bU, _ := untranspose(b)
+	if rv, ok := bU.(RawVectorer); ok {
+		bmat = rv.RawVector()
+		if v != b {
+			v.checkOverlap(bmat)
+		}
+	} else {
+		fast = false
 	}
 
-	a, trans := untranspose(a)
-	ar, ac := a.Dims()
 	v.reuseAs(r)
 	var restore func()
-	if v == a {
-		v, restore = v.isolatedWorkspace(a.(*VecDense))
+	if v == aU {
+		v, restore = v.isolatedWorkspace(aU.(*VecDense))
 		defer restore()
 	} else if v == b {
 		v, restore = v.isolatedWorkspace(b)
 		defer restore()
 	}
 
-	switch a := a.(type) {
+	// TODO(kortschak): Improve the non-fast paths.
+	switch aU := aU.(type) {
 	case *VecDense:
-		if v != a {
-			v.checkOverlap(a.mat)
+		if v != aU {
+			v.checkOverlap(aU.mat)
 		}
 
-		if a.Len() == 1 {
+		if aU.Len() == 1 {
 			// {1,1} x {1,n}
```

Comment:
`{1,1} x {n,1}`?

Reply:
Interesting. If columns of A is 1, rows of b must also be 1. Now we specify that `VecDense` is a column vectore, which means that here we cannot have `{1,1} x {1,n}` unless n is 1.

/cc @btracey

Reply:
I'm tightening this up, and as a second commit removing this case since it doesn't improve much over the fall through at the bottom.